docs/merkletree/treetex/main.go

```diff
@@ -0,0 +1,236 @@
+	maxLen = 64
+)
+
+var (
+	treeSize  = flag.Int64("tree_size", 23, "Size of tree to produce")
+	inclusion = flag.Int64("inclusion", -1, "Leaf index to show inclusion proof")
+
+	// nInfo holds nodeInfo data for the tree.
+	nInfo = make(map[string]nodeInfo)
+)
+
+// nodeInfo represents the style to be applied to a tree node.
+type nodeInfo struct {
+	incProof    bool
+	incPath     bool
+	target      bool
+	perfectRoot bool
+	ephemeral   bool
+	leaf        bool
+}
+
+// String returns a string containing Forest attributes suitable for
+// rendering the node, given its type.
+func (n nodeInfo) String() string {
+	attr := make([]string, 0, 4)
+
+	// Figure out which colour to fill with:
+	fill := "white"
+	if n.perfectRoot {
+		attr = append(attr, "line width=4pt")
+	}
+	if n.incProof {
+		fill = "inclusion"
+		if n.ephemeral {
+			fill = "inclusion_ephemeral"
+			attr = append(attr, "draw, dotted")
+		}
+	}
+	if n.target {
+		fill = "target"
+	}
+	if n.incPath {
+		fill = "target_path"
+	}
+	attr = append(attr, "fill="+fill)
+
+	if !n.ephemeral {
+		attr = append(attr, "draw")
+	}
+	if !n.leaf {
+		attr = append(attr, "circle, minimum size=3em")
+	} else {
+		attr = append(attr, "minimum size=1.5em")
+	}
+	return strings.Join(attr, ", ")
+}
+
+// modifyNodeInfo applies f to the nodeInfo associated with node k.
+func modifyNodeInfo(k string, f func(*nodeInfo)) {
+	n, ok := nInfo[k]
+	if !ok {
+		n = nodeInfo{}
+	}
+	f(&n)
+	nInfo[k] = n
+}
+
+// perfect renders a perfect subtree.
+func perfect(prefix string, height, index int64) {
+	perfectInner(prefix, height, index, true)
+}
+
+// drawLeaf emits TeX code to render a leaf.
+func drawLeaf(prefix string, index int64) {
+	a := nInfo[nodeKey(0, index)]
+	fmt.Printf("%s [%d, %s, tier=leaf]\n", prefix, index, a.String())
+}
+
+// openInnerNode renders TeX code to open an internal node.
+// The caller may emit any number of child nodes before calling the returned
+// func to clode the node.
+// Returns a func to be called to close the node.
+//
+func openInnerNode(prefix string, height, index int64) func() {
+	attr := nInfo[nodeKey(height, index)].String()
+	fmt.Printf("%s [%d.%d, %s, tier=%d\n", prefix, height, index, attr, height-1)
+	return func() { fmt.Printf("%s ]\n", prefix) }
+}
+
+// perfectInner renders the nodes of a perfect internal subtree.
+func perfectInner(prefix string, height, index int64, top bool) {
+	nk := nodeKey(height, index)
+	modifyNodeInfo(nk, func(n *nodeInfo) {
+		n.leaf = height == 0
+		n.perfectRoot = top
+	})
+
+	if height == 0 {
+		drawLeaf(prefix, index)
+		return
+	}
+	c := openInnerNode(prefix, height, index)
+	childIndex := index << 1
+	perfectInner(prefix+" ", height-1, childIndex, false)
+	perfectInner(prefix+" ", height-1, childIndex+1, false)
+	c()
+}
+
+// renderTree renders a tree node and recurses if necessary.
+func renderTree(prefix string, treeSize, index int64) {
+	height := int64(bits.Len64(uint64(treeSize)) - 1)
```

Comment:
Sorry for being picky, I think this line could go 3 lines below, and then the `if` could just check `treeSize <= 0`.